Add block-restricted exodus output

GiudGiud opened the pull request:
refs #30317

GiudGiud committed:
Add block restriction to oversampled output
closes #30317

GiudGiud committed:
Add test for block restricted output
refs #30317

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2814358/), step Clang format on 16caf4f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30318/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30318/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format f5487782f932b1d218cd61807b5b5d06e7227db7`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2814439/), step Docs: sync website on b1cafbe wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30318/site)

This comment will be updated on new commits.

GiudGiud committed:
Remove global Params block restrictions:
- they were not doing anything
- causing a diff on material property output
raised #30320 for the problem

GiudGiud commented on framework/src/outputs/OversampleOutput.C:
```diff
@@ -292,6 +289,22 @@ OversampleOutput::cloneMesh()
+
+    // Remove isolated nodes
+    _cloned_mesh_ptr->getMesh().allow_renumbering(false);
+    _cloned_mesh_ptr->getMesh().prepare_for_use(/*skip_renumber*/ true);
```
```suggestion
    _cloned_mesh_ptr->getMesh().prepare_for_use();
```

GiudGiud committed:
Remove deprecated function call

GiudGiud commented:
SAM failures are legit. I don't have access to Sam out of the country so this will be delayed a week at least

GiudGiud committed:
Rely on advanced output for hide/show & other feature support

GiudGiud committed:
Adapt test for new oversampled output capabilities

GiudGiud committed:
Add distributed mesh support by serializing

GiudGiud committed:
Rename OversampleOutput to SampledOutput

GiudGiud committed:
Rename attributes and clean up logic in sampled output

GiudGiud committed:
Add variable block restriction to prevent sampling increasing their domain of definition

GiudGiud committed:
Add a serial and a parallel mode for sampled output
- rely on the copy-partitioner

GiudGiud committed:
Only allow distributed sampled output with elemental variables, and without a warning on
constant monomials
This is all we can guarantee right now

GiudGiud committed:
Output pids on initial as it could crash right at the beginning

GiudGiud committed:
Allow block in GlobalParams to be used without triggering sampled output
by renaming the block output parameter

GiudGiud committed:
Use explicit syntax instead of shorthand debug option in tests existing before
execution schedule change in debug option

GiudGiud committed:
Adapt tests to sampled output improvements:
- aux data now output

GiudGiud committed:
Enable distributed testing of serialized block-restricted output
Re-organize mesh preparation calls to work for this case
Serialize sampling mesh too

moosebuild commented:
Job [Conda (ARM Mac)](https://civet.inl.gov/job/2927074/) on 6bc3f05 : invalidated by @GiudGiud

unrelated?

GiudGiud commented:
@lindsayad should be good now!

lindsayad reviewed (changes requested):
It's unfortunate that github is showing `SampledOutput.C` as a new file. I see that you had a commit where the move was recognized. Wtf github

Will start with reviewing `SampledOutput.C` tomorrow

GiudGiud commented on framework/include/outputs/SampledOutput.h:
```diff
@@ -81,23 +84,37 @@ class OversampleOutput : public AdvancedOutput
    */
   void cloneMesh();
 
+  /// Used to decide which variable is output as a nodal variable for oversampling purposes
+  bool isSampledAsNodal(const FEType & fe_type) const;
```
prefer isSampledAtNodes ?
it's literally what the class does, we are not using the GenericProjector to decide sampling points

GiudGiud replied:
> but I don't know if it's good in this generic class

I think isSampledAtNodXYZ is good in this class because this class imposes the sampling method as well. See the updateSample routine

GiudGiud replied:
changed

lindsayad commented on framework/src/outputs/SampledOutput.C:
```diff
@@ -0,0 +1,499 @@
+
+  // Create the new EquationSystems
+  _oversample_es = std::make_unique<EquationSystems>(_mesh_ptr->getMesh());
+  _es_ptr = _oversample_es.get();
```
Oh yea the latter is from the base class and used in derived classes

GiudGiud commented on framework/src/outputs/SampledOutput.C:
```diff
@@ -0,0 +1,499 @@
+  params.addParam<unsigned int>("refinements",
+                                0,
+                                "Number of uniform refinements for oversampling "
+                                "(refinement levels beyond any uniform "
+                                "refinements)");
```
yes

GiudGiud commented on framework/src/outputs/SampledOutput.C:
```diff
@@ -0,0 +1,499 @@
+  params.addDeprecatedParam<bool>("append_oversample",
+                                  false,
+                                  "Append '_oversample' to the output file base",
+                                  "This parameter is no longer operational, to append "
```
I ll reword this

GiudGiud commented on framework/src/outputs/SampledOutput.C:
```diff
@@ -0,0 +1,499 @@
+  if (_using_external_sampling_file)
+    _cloned_mesh_ptr->prepare(/*mesh to clone*/ nullptr);
+  else if (_serialize && isParamValid("sampling_blocks"))
+    // TODO: constraints have not been initialized?
```
it s a good question. If i just use prepare() here I get some complaint about constraint

GiudGiud commented on framework/src/outputs/SampledOutput.C:
```diff
@@ -0,0 +1,499 @@
+    // we want to avoid re-partitioning, as we will serialize anyway
+    _cloned_mesh_ptr->getMesh().skip_partitioning(true);
+
+  // Make sure that the mesh pointer points to the newly cloned mesh
+  _mesh_ptr = _cloned_mesh_ptr.get();
```
will do

GiudGiud replied:
nevermind. we need a place to store the sampling mesh.

renamed the attribute, added a docstring

GiudGiud commented on modules/combined/test/tests/adaptive_timestepping/adapt_tstep_function_change.i:
```diff
@@ -9,7 +9,6 @@
   displacements = 'disp_x disp_y disp_z'
   order = FIRST
   family = LAGRANGE
-  block = 1
```
I initially named the sampling parameter block. But then got hit with everywhere it is used in the GlobalParams...

I dont think block fits well in global params

lindsayad replied:
A lot of people use it though. It has cropped up for discussion quite a bit in https://github.com/idaholab/moose/pull/30300. I'm fine with your explanation and removal

GiudGiud commented on modules/solid_mechanics/test/tests/anisotropic_patch/anisotropic_patch_test.i:
```diff
@@ -23,7 +23,6 @@
 
 [GlobalParams]
   displacements = 'disp_x disp_y disp_z'
-  block = '1 2 3 4 5 6 7'
```
see explanation above

GiudGiud commented on test/tests/misc/debug_pid_aux/debug_pid_aux.i:
```diff
@@ -11,8 +11,15 @@
 
-[Debug]
-  output_process_domains = true
+[AuxVariables]
+  [pid]
+    family = MONOMIAL
+    order = CONSTANT
+    [AuxKernel]
+      type = ProcessorIDAux
+      execute_on = 'TIMESTEP_END'
+    []
+  []
```
I added the EXEC_INITIAL to the Debug syntax. This test was golded with the old debug syntax. To avoid regolding, we re using the auxkernel directly now

GiudGiud committed:
Address Alex's review:
- better parameter description
- use sampling/sampled instead of oversampling/sampled
- delete second mesh ptr attribute

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

GiudGiud commented on framework/src/outputs/SampledOutput.C:
```diff
@@ -0,0 +1,499 @@
+
+  // This should be called after changing the mesh (block restriction for example)
+  if (_change_position || (_refinements > 0) || isParamValid("sampling_blocks"))
+    _cloned_mesh_ptr->meshChanged();
```
I think we can just delete this attribute and live with _mesh_ptr

GiudGiud replied:
renamed to _sampling_ . Could also be sampled_

GiudGiud commented on modules/phase_field/test/tests/KKS_system/kks_example.i:
```diff
@@ -19,33 +19,33 @@
 [Variables]
   # order parameter
-  [./eta]
+  [eta]
```
this is an old PR

GiudGiud commented on test/tests/partitioners/single_rank_partitioner/single_rank_partitioner.i:
```diff
@@ -44,8 +44,15 @@
   petsc_options_value = 'hypre boomeramg'
 []
 
-[Debug]
```
it's really better that way. output_process-domains on timestep_end after it had already crashed was not useful!

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2942001/), step Clang format on 8cb498c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30318/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30318/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0fe928d85dc6585ae41c4d42f0d68c02442cbd58`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2942089/) on a0469a6 : invalidated by @lindsayad

GiudGiud commented:
Thanks for the review!
